docs/adr/adr-004-state-interaction.md

```diff
@@ -0,0 +1,189 @@
+
+Both celestia **light** and **full** nodes will run `StateService`. 
+`StateService` will be responsible for managing the RPC endpoints provided by `StateAccessor`.
+
+`StateAccessor` will be an interface for interacting with celestia-core in order to retrieve account-related information
+as well as submit transactions.
+
+### `StateService`
+
+`StateService` will expose several higher-level RPC endpoints for users to access the methods provided by the 
+`StateAccessor`.
+
+```go
+type StateService struct {
+    accessor StateAccessor
+}
+``` 
+
+### `StateAccessor`
+
+`StateAccessor` is defined by the two methods listed below and may be expanded in the future to include more methods 
+related to accounts and transactions in later iterations.
+
+```go
+type StateAccessor interface {
+    func AccountBalance(context.Context, Account) (Balance, error)
+    func SubmitTx(context.Context, Tx) error
+}
+```
+
+`StateAccessor` will have 3 separate implementations under the hood: 
+1. **CORE**: RPC connection with a celestia-core endpoint handed to the node upon initialisation
+(*required for this iteration*) 
+2. **P2P**: discovery of a **bridge** node, and then establish direct RPC connection (*nice-to-have for this iteration*)
+3. **LOCAL**: eventually, **full** nodes will be able to provide state to the network, and can execute the
+
+### `CoreAccess`
+
+`CoreAccess` will be the direct RPC connection implementation of `StateAccessor`. It will use the [lens implementation of ChainClient](https://github.com/strangelove-ventures/lens/blob/main/client/chain_client.go#L23)
+under the hood. `lens.ChainClient` provides a nice wrapper around the Cosmos-SDK APIs. 
+
+Upon node initialisation, the node will receive the endpoint of a trusted celestia-core node. The constructor for 
+`CoreAccess` will construct a new instance of `CoreAccess` with the `lens.ChainClient` pointing at the given 
+celestia-core endpoint.
+
+```go
+type CoreAccess struct {
+    cc *lens.ChainClient
+}
+
+func NewCoreAccess(cc *lens.ChainClient) (*CoreAccess, error) {
+	return &CoreAccess{
+        cc: cc,		
+    }, nil   
+}
+
+// where Account wraps account information and Balance 
+func (ca *CoreAccess) AccountBalance(ctx context.Context, acct Account) (Balance, error) {
+	coins, err := da.cc.QueryBalanceWithAddress(acct.address)
+	if err != nil {
+		return err
+    }   
+    // returning first index as we only care for account's balance on celestia chain	
+	return coins[0], nil 
+}
+
+// where Tx implements sdk.Msg
+func (ca *CoreAccess) SubmitTx(ctx context.Context, tx Tx) (TxResponse, error) {
+	return da.cc.SendMsg(ctx, tx)
+}
+```
+
+### `P2PAccess`
+
+An idea for how `P2PAccess` could be implemented can be found below.
+
+```go
+type P2PAccess struct {
+	host libhost.Host
+    // P2PAccess will still use lens.ChainClient's methods
+	// under the hood, but will use libp2p transport instead
+	// of RPC communication with the Core endpoint.
+   cc  *lens.ChainClient 
```

Comment:
Pls, provide proof that this is feasible. This is the part of the point which I said will be ugly and has multiple caveats that need to be double-checked before doing the implementation.

Reply:
Otherwise, we shouldn't use the lens for the p2p approach and only for RPC.

Reply:
Ok, so I checked and lens [uses](https://github.com/strangelove-ventures/lens/blob/main/client/chain_client.go#L27) tm's RPCClient or `core.Client` in our codebase, so to make this happen we would need to implement `core.Client` over P2P.

Reply:
Which I think would be better. We don't need to fully implement all methods either, just those we need as we go along

Reply:
> Which I think would be better

Any arguments supporting this point?

Reply:
Note that, I am not completely against this way, I just don't see we should go this way. If we merge this point as part of the ADR, everyone on the team should be aligned with this trajectory so pls explain.

Reply:
I understand that you want to do less work by reusing lens, but I am afraid about this being even more work and overhead than just using p2p streams directly. Devil is always in detail, especially when you are trying to connect two different worlds into something workable.

Reply:
Okay looking into it now, I think it would be easier to just not use the chain client in p2p, but that would require us doing a lot of other work from scratch instead of shoehorning things around the ChainClient implementation. I guess I just liked the fact all the methods we need are already implemented, including keyring management parts.

Reply:
Okay.

For this, I bet, we would need to implement a custom key management api.
> including keyring management parts.

Reply:
Resolved by my side. Resolve it via the button as an ack that you read this.